File: Combination Sum II.cpp
class Solution {
public:
    void combinationSum2(vector<int>& candidates, int start, int target, vector<int>& solution, vector<vector<int>>& result) {
        if (target == 0) {
            result.push_back(solution);
            return;
        }
        if (start >= candidates.size())
            return;
        if (candidates[start] > target)
            return;
        int count = 0;
        while (start + count < candidates.size() && candidates[start] == candidates[start + count])
            count++;
        for (int i = 0; i <= count; i++) {
            combinationSum2(candidates, start + count, target - i * candidates[start], solution, result);
            if (i != count)
                solution.push_back(candidates[start]);
        }
        solution.resize(solution.size() - count);
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> solution;
        sort(candidates.begin(), candidates.end());
        combinationSum2(candidates, 0, target, solution, result);
        return result;
    }
};

File: Combination Sum II.java
public class Solution {
    private void find(List<Integer> chosen, List<Integer> num, int target, List<List<Integer>> result) {
        if (target == 0) {
            List<Integer> clone = new ArrayList<Integer>(chosen);
	        result.add(clone);
            return;
        }
        if (num.isEmpty())
            return;
        
        List<Integer> candidates = new ArrayList<Integer>();
        int sum = 0;
        for (Integer i : num) {
            if (i > target)
                break;
            candidates.add(i);
            sum += i;
        }
        if (sum < target)
            return;
            
        Integer biggest = candidates.get(candidates.size() - 1);
        candidates.remove(candidates.size() - 1);
        // select one biggest at least
        chosen.add(0, biggest);
        find(chosen, candidates, target - biggest, result);
        
        // selet no biggest at all
        chosen.remove(0);
        while (!candidates.isEmpty() && candidates.get(candidates.size() - 1) == biggest)
            candidates.remove(candidates.size() - 1);
        find(chosen, candidates, target, result);
    }
    public List<List<Integer>> combinationSum2(int[] num, int target) {
        Arrays.sort(num);
        List<Integer> list_num = new ArrayList<Integer>();
        for (int i = 0; i < num.length; i++)
            list_num.add(num[i]);
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> chosen = new ArrayList<Integer>();
        find(chosen, list_num, target, result);
        return result;
    }
}